course5/homework04/homework04.cc

```diff
@@ -0,0 +1,122 @@
+#include <iostream>
+#include <cstdio>
+#include <cstdlib>
+#include <stdexcept>
+#include <string>
+
+/***
+Define a “Stream” class, that allows for easy printing of values to a file
+- Let the class manage the file: ie open the file in constructor and close it in destructor.
+- Overload operator<< for the class, so that it can be used to print some types.
+- Define a global variable of the “Stream” type, call it, say, “Cout”, that will write to stdout.
+***/
+
+class Stream
+{
+public:
+	// Constructors
+	Stream() // default
+	{
+		throw std::runtime_error("Stream must be constructed with a filename: Stream('filename.txt')");
+	}
+	Stream(const Stream&) // copy
+	{
+		throw std::runtime_error("Stream object cannot be copied");
+	}
+	Stream(Stream&& src) // move
+	{
+		fname = src.fname;
+		fp = src.fp;
```

Comment:
I wrote some tests, initially like this:
`Stream s2 = std::move(Stream("another.txt"));`
because I wanted an actual anonymous object that wasn't assigned to any variables.

But the compiler kept spitting out `warning: moving a temporary object prevents copy elision [-Wpessimizing-move]`, and the "another.txt" file would never be written to. That made it sound like the compiler could NOT do the "copy elision" optimization -- but the copy constructor WAS being called. What was going on?